presto-native-execution/presto_cpp/main/operators/StorageBasedExchangeSource.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/common/file/File.h"
+#include "velox/common/file/FileSystems.h"
+#include "velox/core/PlanNode.h"
+#include "velox/exec/Exchange.h"
+#include "velox/exec/Operator.h"
+#include "velox/serializers/PrestoSerializer.h"
+
+namespace facebook::presto::operators {
+
+class StorageBasedExchangeSource : public velox::exec::ExchangeSource {
+ public:
+  StorageBasedExchangeSource(
+      const std::string& taskId,
+      int destination,
+      std::shared_ptr<velox::exec::ExchangeQueue> queue,
+      const std::vector<std::string>& inputFolders,
```

Comment:
Just to clarify, do we have one folder per task?

Reply:
Yes, one folder per producer tasks. And the consumer will get the list of all of the produced folders.